### DISCUSSION TITLE
Unable to find entrypoint after deploying in staging.

### DISCUSSION THREAD

**Discussion by rachitpant** [ElMassimo/vite_ruby#197] [Q&A] [answered]
Hello, 
First i want to thank you  for bringing the power of Vite to the rails community. Now the actual issue.

I migrated a moderately large rails/vue project from webpacker to vite. Its working perfectly fine locally. The problem happens when i try to deploy to staging environment using capistrano. 
The one major difference between local and staging environment is that we have a base location ("/aasp") under which everthing is deployed.

Here is what i see in staging.log.
`ActionView::Template::Error (Vite Ruby can't find entrypoints/application.js in ../../../../../manifest.json or ../../../../../manifest-assets.json.`

However the manifest file is present in the public directory , and it does have an entry for the entrypoint application.js
Here is the structure of the deployment directory on the server.
<img width="647" alt="image" src="https://user-images.githubusercontent.com/1474415/159303502-c1b9c285-4e95-45e2-94b2-7858dad5ca66.png">

Also here is part of the manifest file
```
{
  "entrypoints/application.js": {
    "file": "assets/application.b3572247.js",
    "src": "entrypoints/application.js",
    "isEntry": true,
    "imports": [
      "_vendor.9de18915.js"
    ],
    "dynamicImports": [
      "components/layout/Loading.vue",
      "components/layout/Header.vue",
      "components/layout/Nav.vue",
      "components/layout/Footer.vue",
      "pages/Home.vue",
      "pages/admin/GrantList.vue",
```
Here is the assets directory where the bundled `assets/application.b3572247.js` is present.
<img width="503" alt="Screen Shot 2022-03-21 at 12 15 09 PM" src="https://user-images.githubusercontent.com/1474415/159304225-66695bc5-301f-4cec-9a63-2698e34a2ed0.png">

When I first got this error , I realized our Apache and passenger are configured so that assets live in a shared asset directory, so after deployment I moved all the assets that were originally present in public/assets directory ,  to the shared assets directory and made public/assets a symlink to the shared assets. This can be seen in the Fiilezilla screenshots above. All this also did not help. I just thought i should mention that.

Other than this, the publicOutputDir is "/" in vite.json
```
  "development": {
    "autoBuild": true,
    "publicOutputDir": "vite-dev",
    "port": 3036
  },
  "staging": {
    "publicOutputDir": "/"
  }
}
```
Also since the base URL is different in staging environment.
This is part of my vite.config.js
```
export default defineConfig({
  base: `"${process.env.routerBase}"`,
  plugins: [
```
and I added the folowing line to my `deploy.rb` file.
`set :default_env, {VITE_RUBY_BASE: (ARGV[0] == "production" ? "/" : "/aasp") }`

I know I am so close , especially since i got the "hard" parts working locally , like using VCurrencyField and vite with vue. However getting all this to work in staging is giving a real headache. Any help will be appreciated.

**Comment by ElMassimo** [comment, marked as answer]
Hi Rachit! @rachitpant 

Regarding the first error, you will want to ensure that [`ViteRuby.config.manifest_path`](https://github.com/ElMassimo/vite_ruby/blob/main/vite_ruby/lib/vite_ruby/config.rb#L18) exists, which you can check manually by starting a console in the server.

From [the error](https://github.com/ElMassimo/vite_ruby/blob/main/vite_ruby/lib/vite_ruby/missing_entrypoint_error.rb#L14), it seems that the assets dir is outside the root of the Rails app, which is very unusual.

--- 

You should not configure `base` manually in Vite, instead use the [`base` config option](https://vite-ruby.netlify.app/config/#base):

```json
"base": "/aasp"
```

I'd advise not modifying `publicOutputDir` and keep the `vite` default. If you do, it should be `"publicOutputDir": ""` instead of containing a slash (not a valid dir name).

**Reply by rachitpant** [reply]
> Hi Rachit! @rachitpant
> 
> Regarding the first error, you will want to ensure that [`ViteRuby.config.manifest_path`](https://github.com/ElMassimo/vite_ruby/blob/main/vite_ruby/lib/vite_ruby/config.rb#L18) exists, which you can check manually by starting a console in the server.
> 
> From [the error](https://github.com/ElMassimo/vite_ruby/blob/main/vite_ruby/lib/vite_ruby/missing_entrypoint_error.rb#L14), it seems that the assets dir is outside the root of the Rails app, which is very unusual.

Here is the manifest path @ElMassimo 
```
[rpant@ruby1 20220321151625]$ RAILS_ENV=staging bin/rails c
Warning: NLS_LANG is not set. fallback to US7ASCII.
Loading staging environment (Rails 6.1.4.4)
2.7.2 :001 > ViteRuby.config.manifest_path
 => #<Pathname:/manifest.json> 
2.7.2 :002 > ViteRuby.config.manifest_path.relative?
 => false 
2.7.2 :003 > ViteRuby.config.manifest_path.absolute?
 => true 
```


> You should not configure `base` manually in Vite, instead use the [`base` config option](https://vite-ruby.netlify.app/config/#base):
> 
> ```json
> "base": "/aasp"
> ```

Do you mean removing the configuration from `vite.config.json` and adding the base configuration only in `vite.json`. Also should i keep the following in deploy.rb
```set :default_env, {VITE_RUBY_BASE: (ARGV[0] == "production" ? "/" : "/aasp") }```

**Reply by ElMassimo** [reply]
The manifest path does not seem correct, probably because you are using `"publicOutputDir": "/"`, which is not a valid config.

I'd avoid using `VITE_RUBY_BASE`, and instead set `"base": "/aasp"` for staging (if that's what you are using).

**Reply by rachitpant** [reply]
@ElMassimo Thank you for the quick help. I got the application running by making these changes and then deploying. I left the base url in `vite.config.js` , but removed it from the deploy script like you suggested , also corrected the ```publicOutputDir```.

I had another minor problem related to favicon which i had commented out for now , but on uncommenting I am getting an error on the server.
The following in `application.html.erb` works fine locally
`<%= favicon_link_tag vite_asset_path "images/favicon.ico" %>`

However on the server i get an error.
```
F, [2022-03-21T13:26:57.076720 #6342] FATAL -- :   
ActionView::Template::Error (Vite Ruby can't find images/favicon.ico in public/manifest.json or public/manifest-assets.json.
```

**Reply by ElMassimo** [reply]
Check that `ViteRuby.config.assets_manifest_path` exists, and `images/favicon.ico` is in `manifest-assets.json`.

**Reply by rachitpant** [reply]
@ElMassimo For some reason there is no` maniest.json` or `manifest-assets.json` in the public directory although `ViteRuby.config.assets_manifest_path`  does exist.
```
Loading staging environment (Rails 6.1.4.4)
2.7.2 :001 > ViteRuby.config.manifest_path
 => #<Pathname:/var/www/aasp/releases/20220321174332/public/manifest.json> 
```
The public directory on the server is an exact copy of the local public directory, plus an additional asset folder.
<img width="621" alt="image" src="https://user-images.githubusercontent.com/1474415/159337447-950ff6b2-befd-4576-83f7-89ac4b748b91.png">

Just doesn't make any sense. Plus my happiness of getting the app to work in staging was short lived. I undid the changes related to getting the favicon to show , git forced , and now even after undoing all this , i can't get the application to run on staging.

**Reply by rachitpant** [reply]
Also wondering if the error continues to show because vite build never really happened. The only file that was changed ( and undid ) was `application.html.erb` , should that trigger a new build.

 ```
00:24 deploy:assets:precompile
     01 /usr/local/rvm/bin/rvm 2.7.2 do bundle exec rake assets:precompile
      01 Warning: NLS_LANG is not set. fallback to US7ASCII.
      01 npx: installed 1 in 1.541s
      01 yarn install v1.15.2
      01 [1/4] Resolving packages...
      01 success Already up-to-date.
      01 Done in 0.41s.
      01 Skipping vite build. Watched files have not changed since the last build at 2022-03-21 13:23:34
      01 yarn install v1.15.2
      01 [1/4] Resolving packages...
      01 success Already up-to-date.
      01 Done in 0.41s.
```

**Comment by rachitpant** [comment]
@ElMassimo I ran `RAILS_ENV=staging bin/rake vite:clobber` and then triggered a rebuild and now everything is working fine again. 

There are two things that seem like an issue to me during deployment :
 1) If the rails assets directory is an existing symlink , the vite build is not placing the assets in the actual location , but instead its recreating an assets folder and placing assets there.
2) The actual content of the public directory ( in addition to assets ) is either the manifest files ( manifest.json and manifest-assets.json ) or the rails application public folder content. I am not sure when one happens vs the other , but i have never been able to get both the orignal public directory contents and manifest files within the public directory of the server deployment.
